cranelift/isle/isle/src/serialize.rs

```diff
@@ -0,0 +1,823 @@
+    /// first in `order`. Return the number of rules which are valid with
+    /// respect to priority, as well as whether any rules matched the query at
+    /// all. No ordering is guaranteed within either partition, which allows
+    /// this function to run in linear time. That's fine because later we'll
+    /// recursively sort both partitions.
+    fn partition(self, rules: &RuleSet, order: &mut [usize]) -> PartitionResults {
+        let matching = partition_in_place(order, |&idx| {
+            let rule = &rules.rules[idx];
+            match self {
+                HasControlFlow::Match(binding_id) => rule.get_constraint(binding_id).is_some(),
+                HasControlFlow::Equal(x, y) => rule.equals.in_same_set(x, y),
+                HasControlFlow::Loop(binding_id) => rule.iterators.contains(&binding_id),
+            }
+        });
+        PartitionResults {
+            any_matched: matching > 0,
+            valid: respect_priority(rules, order, matching),
+        }
+    }
+}
+
+/// As we proceed through sorting a term's rules, the term's binding sites move
+/// through this sequence of states. This state machine helps us avoid doing
+/// the same thing with a binding site more than once in any subtree.
+#[derive(Clone, Copy, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+enum BindingState {
+    /// Initially, all binding sites are unavailable for evaluation except for
+    /// top-level arguments, constants, and similar.
+    #[default]
+    Unavailable,
+    /// As more binding sites become available, it becomes possible to evaluate
+    /// bindings which depend on those sites.
+    Available,
+    /// Once we've decided a binding is needed in order to make progress in
+    /// matching, we emit a let-binding for it. We shouldn't evaluate it a
+    /// second time, if possible.
+    Emitted,
+    /// We can only match a constraint against a binding site if we can emit it
+    /// first. Afterward, we should not try to match a constraint against that
+    /// site again in the same subtree.
+    Matched,
+}
+
+/// A sort key used to order control-flow candidates in `best_control_flow`.
+#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+struct Score {
+    // We prefer to match as many rules at once as possible.
+    count: usize,
+    // Break ties by preferring bindings we've already emitted.
+    state: BindingState,
+}
+
+/// A rule filter ([HasControlFlow]), plus temporary storage for the sort
+/// key used in `best_control_flow` to order these candidates. Keeping the
+/// temporary storage here lets us avoid repeated heap allocations.
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+struct Candidate {
+    score: Score,
+    // Last resort tie-breaker: defer to HasControlFlow order, but prefer
+    // control-flow that sorts earlier.
+    kind: Reverse<HasControlFlow>,
+}
+
+impl Candidate {
+    /// Construct a candidate where the sort key is not set. The sort key will
```

Comment:
```suggestion
    /// Construct a candidate where the sort key is not set. The score will
```

Reply:
There are more references to fields named `score` as `sort key`, so maybe this isn't a big deal :)

Reply:
Or maybe it means I should make changes in more places than this for consistency :sweat_smile: 

I was hoping the doc comment for `Score` describing it as "A sort key ..." would be clear enough about the two terms being interchangeable. But then, if they're interchangeable, then why not pick one and stick with it?